### PR TITLE
added the Lambda for getUserById

### PR DESCRIPTION
**Pull request by syedautherabbas**

### CODE REVIEW

**Review by antoniolofiego — CHANGES_REQUESTED** [review]
Two main comments:

- Can you add a few comments and a docstring?
- Do we have a test for this one? You can use `moto` for mocking boto3 resources.

**Comment by what-name** [comment]
- Please manually add CORS headers in the Lambda response

**Comment by syedautherabbas** [comment]
Added the docs and the cors header.

File: API/getUserById/getUserById.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    )
+
+    logger = logging.getLogger()
+
+    git_user=""
+    resp={}
+    statusCode=404
+    
+    #checking if the lambda is called by an API gateway and has query params in the call
+    if "queryStringParameters" in event:
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:33]: I don't think it is necessary to check for this, as this won't be an user-facing API endpoint, so we already know where this will be used and called from.

File: API/getUserById/getUserById.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    	#checking if the query params from the API call has a username
+        if "username" in event["queryStringParameters"]:
+            git_user=event["queryStringParameters"]['username']
+    if git_user=="":
+        resp['message']='User Not found'
+    else:   
+        try:
+        	#checking for database name from env variables on the lambda
+            if 'databaseName' in os.environ:
+                table = dynamodb.Table(os.environ['databaseName'])
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:43]: We should just start at this stage. Calling the table that we need to call in a try-catch block and handle errors appropriately in the catch block

File: API/getUserById/getUserById.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                #querying the dynamoDb with the primary key github_username
+                response = table.get_item(Key={'github_username': git_user})
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:45]: What would be the `response` variable if the item is not found?

File: API/getUserById/getUserById.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                if "Item" in  response:
+                    resp= response['Item']
+                    statusCode=200
+                #If there is no 'Item' found in response it means that there was no data for that primary key and user does not exist    
+                else:
+                     statusCode=404   
+                     resp['message']='User Not found'
+                     logger.error('User '+str(git_user)+' not found')
+            else:    
+                statusCode=500
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:56]: Again, kind of unnecessary as we have well defined tables that this function works with.

File: API/getUserById/getUserById.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                #This means we need a valid table name to query and get the data. 
+                resp['message']="Server error, table not configured"
+                logger.error('Server error, table not configured')
+
+        except ClientError as e:
+            print(e.response['Error']['Message'])
+            resp['message']=e.response['Error']['Message']
+            logger.error(e.response['Error']['Message'])
+
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:65]: Good logging ideas

File: API/getUserById/getUserById.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        except ClientError as e:
+            print(e.response['Error']['Message'])
+            resp['message']=e.response['Error']['Message']
+            logger.error(e.response['Error']['Message'])
+        
+           
+    
+    	
+    final_response= {
+    	'headers': {
```
Comment (antoniolofiego) [inline API/getUserById/getUserById.py:70]: Can directly return this, no need to instantiate a variable that will be not used anywhere else

**Review by antoniolofiego — CHANGES_REQUESTED** [review]
Please check the comments. We can speed up the query by directly acting on the table and handling errors appropriately.
Also we need a testing suite for the edge cases that I highlighted. Refer to me or @what-name for testing with `moto` and `pytest` if you have not used it before.

File: API/getUserById/getUserById.py (hunk 2 above)
Comment (what-name) [reply API/getUserById/getUserById.py:43]: Also, `databaseName` will for sure be in the environment variable, no need to check for it IMO. If it's not there, the function will fail anyways. If the env variable does not correlate with an actual table, it will fail anyways. Kind of unnecessary to check for this variable existing.
We gotta have unit tests that check this as well (existing vs not existing env variable)

**Comment by syedautherabbas** [comment]
addressing testing requests by using moto for dynamoDb mocking, updated naming conventions in line with pytest.

**Comment by what-name** [comment]
- @syedautherabbas After #23 gets merged, can you move the `API/get_user_by_Id/` folder into the `./SAM/API/Lambda/get_user_by_Id/` please. I will edit the `api.template.yml` file accordingly after that.
- The pytest is currently testing itself instead of importing the Lambda function. The `test_get_user_by_ID` function needs to use the `lambda_handler` from the `get_user_by_Id.py` file. Furthermore, there needs to be a `__main__` initializer that runs the test function on the command `python3 testxyz.py`.
